pygen/pygen_src/riscv_instr_sequence.py

```diff
@@ -1,89 +1,191 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+"""
+import logging
+import random
+from collections import defaultdict
+from pygen_src.riscv_instr_stream import riscv_rand_instr_stream
+from pygen_src.riscv_instr_gen_config import cfg
+from pygen_src.riscv_instr_pkg import pkg_ins
+
+
+class riscv_instr_sequence:
+
+    def __init__(self):
+        self.instr_cnt = 0
+        self.instr_stream = riscv_rand_instr_stream()
+        self.is_main_program = 0
+        self.is_debug_program = 0
+        self.label_name = ""
+        self.instr_string_list = []  # Save the instruction list
+        self.program_stack_len = 0  # Stack space allocated for this program
+        self.directed_instr = []    # List of all directed instruction stream
+        self.illegal_instr_pct = 0  # Percentage of illegal instructions
+        self.hint_instr_pct = 0     # Percentage of hint instructions
+        self.branch_idx = [None] * 30
+
+    def gen_instr(self, is_main_program, no_branch = 1):
+        self.is_main_program = is_main_program
+        self.instr_stream.initialize_instr_list(self.instr_cnt)
+        logging.info("Start generating %d instruction" % len(self.instr_stream.instr_list))
+        self.instr_stream.gen_instr(no_branch = no_branch, no_load_store = 1,
+                                    is_debug_program = self.is_debug_program)
+
+        if not is_main_program:
+            self.gen_stack_enter_instr()
+            self.gen_stack_exit_instr()
+
+    # TODO
+    def gen_stack_enter_instr(self):
+        pass
+
+    # TODO
+    def gen_stack_exit_instr(self):
+        pass
+
+    '''
+    ----------------------------------------------------------------------------------------------
+    Instruction post-process
+
+    Post-process is required for branch instructions:
+
+    Need to assign a valid branch target. This is done by picking a random instruction label in
+    this sequence and assigning to the branch instruction. All the non-atomic instructions
+    will have a unique numeric label as the local branch target identifier.
+    The atomic instruction streams don't have labels except for the first instruction. This is
+    to avoid branching into an atomic instruction stream which breaks its atomicy. The
+    definition of an atomic instruction stream here is a sequence of instructions which must be
+    executed in-order.
+    In this sequence, only forward branch is handled. The backward branch target is implemented
+    in a dedicated loop instruction sequence. Randomly choosing a backward branch target could
+    lead to dead loops in the absence of proper loop exiting conditions.
+    ----------------------------------------------------------------------------------------------
+    '''
+
+    def post_process_instr(self):
+        label_idx = 0
+        branch_cnt = 0
+        j = 0
+
+        def def_val():
```

Comment:
Can't we take this `def_val` function out of the `post_process_instr`? What's the purpose of it?

Reply:
Hi @hodjat91 
I have taken branch_target as `defaultdict()` such that it doesn't throw any key error.
So, to initialize the dictionary with a default value i.e 0, I have taken the def_val() function because the first argument in the defaultdict() must be **callable** or **None**. 

I can think of 2 scenarios here
1) Like you said to define the def_val() function outside the post_process() function. or
2) Provide the default value as None instead of 0.

For the second case, we wont need the def_val() function.
What do you think?

Reply:
Thanks for the clarification, @aneels3 . I got it. I think either way should be fine, though I prefer the second option as it avoids using helper function (which might make it a bit confusing).

Reply:
Fixed Now!
Thanks @hodjat91